### PR TITLE
feat(Modal): allow to block focus lock from stealing focus for other type of dialogs

### PR DESCRIPTION
**Pull request by YossiSaadi**
### **PR Type**
Enhancement

https://monday.monday.com/boards/3532714909/pulses/9951664841

___

### **Description**
- Add `useFocusEscapeTargets` hook for managing focus escape targets

- Implement `allowFocusEscapeTo` prop to allow specific elements outside modal focus

- Simplify focus lock whitelist logic with improved target matching

- Add comprehensive test suite for focus escape targets hook

- Clarify `onFocusAttempt` callback documentation and behavior

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Modal Props"] -->|"allowFocusEscapeTo"| B["useFocusEscapeTargets Hook"]
  B -->|"shouldAllowFocusEscape"| C["handleFocusLockWhiteList"]
  C -->|"manages focus"| D["Focus Lock"]
  E["onFocusAttempt Callback"] -->|"custom logic"| C
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>useFocusEscapeTargets.ts</strong><dd><code>Create useFocusEscapeTargets hook for focus management</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts

<ul><li>New hook that creates a checker function to determine if elements <br>should escape focus lock<br> <li> Pre-processes targets into combined CSS selectors and element list for <br>performance<br> <li> Implements <code>shouldAllowFocusEscape</code> callback using <code>closest()</code> for <br>selectors and <code>contains()</code> for elements<br> <li> Supports string selectors, HTMLElement references, and React refs</ul>

</details>

  </td>
  <td><a href="https://github.com/mondaycom/vibe/pull/3143/files#diff-7e588c0f340d38e53f11108e8b73c0ee82ed7a0afb7e95f457e4233a40ec7f8b">+72/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>useFocusEscapeTargets.types.ts</strong><dd><code>Define FocusEscapeTarget type for focus escape targets</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.types.ts

<ul><li>Define <code>FocusEscapeTarget</code> type supporting three target formats<br> <li> Allows CSS selectors, direct HTMLElement references, or React <br>RefObjects</ul>

</details>

  </td>
  <td><a href="https://github.com/mondaycom/vibe/pull/3143/files#diff-baf26eecd4711b4dabe9b0393398588122eee27b67bd3354930cdc2f41427841">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export useFocusEscapeTargets hook and types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/core/src/components/Modal/hooks/useFocusEscapeTargets/index.ts

- Export hook and types from the useFocusEscapeTargets module

</details>

  </td>
  <td><a href="https://github.com/mondaycom/vibe/pull/3143/files#diff-ded555a69197bd60aec3735e0a85fda997493707ff43178a16511e7eeda8e73e">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Modal.tsx</strong><dd><code>Integrate focus escape targets into Modal focus management</code></dd></summary>
<hr>

packages/core/src/components/Modal/Modal/Modal.tsx

<ul><li>Import new <code>useFocusEscapeTargets</code> hook<br> <li> Add <code>allowFocusEscapeTo</code> prop to Modal component destructuring<br> <li> Call <code>useFocusEscapeTargets</code> hook with <code>allowFocusEscapeTo</code> prop<br> <li> Refactor <code>handleFocusLockWhiteList</code> to integrate both <code>onFocusAttempt</code> <br>callback and <code>shouldAllowFocusEscape</code> check<br> <li> Invert logic: return <code>true</code> to manage focus, <code>false</code> to surrender control <br>to element<br> <li> Add clarifying comments about focus-lock behavior</ul>

</details>

  </td>
  <td><a href="https://github.com/mondaycom/vibe/pull/3143/files#diff-ae273d1fbb99fc4e4045d805b44f26f9c8844b9be0a70167269daeba64063141">+17/-7</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>useFocusEscapeTargets.test.ts</strong><dd><code>Add comprehensive test suite for useFocusEscapeTargets</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/core/src/components/Modal/hooks/useFocusEscapeTargets/__tests__/useFocusEscapeTargets.test.ts

<ul><li>Add 136 lines of comprehensive test coverage for useFocusEscapeTargets <br>hook<br> <li> Test scenarios include no targets, CSS selectors, HTMLElement targets, <br>RefObject targets, and mixed target types<br> <li> Verify correct matching behavior including nested elements, direct <br>matches, and edge cases<br> <li> Test null/undefined handling and graceful degradation</ul>

</details>

  </td>
  <td><a href="https://github.com/mondaycom/vibe/pull/3143/files#diff-08881c9be46db0157ccf2b69fed223484e64c0064b90f24417fd925990b2eb5a">+136/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Modal.types.tsx</strong><dd><code>Add allowFocusEscapeTo prop and clarify focus callback docs</code></dd></summary>
<hr>

packages/core/src/components/Modal/Modal/Modal.types.tsx

<ul><li>Add new <code>allowFocusEscapeTo</code> prop to ModalProps interface<br> <li> Expand <code>onFocusAttempt</code> documentation with clarified behavior and <br>example usage<br> <li> Document that <code>onFocusAttempt</code> return value <code>true</code> means focus-lock <br>manages element, <code>false</code> means ignore it<br> <li> Add note recommending <code>allowFocusEscapeTo</code> for simple selector-based <br>focus escape needs</ul>

</details>

  </td>
  <td><a href="https://github.com/mondaycom/vibe/pull/3143/files#diff-f9e80e362c460bed0043a8eb47d2d1a0368bb1b8236f0869c0ee744d8a792f82">+28/-6</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  const { combinedSelector, elements } = useMemo(() => {
+    if (!targets || targets.length === 0) {
+      return { combinedSelector: null, elements: [] };
+    }
+
+    const selectorList: string[] = [];
+    const elementList: HTMLElement[] = [];
+
+    for (const item of targets) {
+      if (typeof item === "string") {
```
Comment (rivka-ungar) [inline packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts:22]:
```suggestion
      if (typeof item === "string" || item instanceof HTMLElement) {
```

File: packages/core/src/components/Modal/Modal/Modal.types.tsx [hunk 2]
```diff
@@ -44,18 +45,39 @@ export interface ModalProps extends VibeComponentProps {
...
-   * - `true` to allow normal flow focus.
-   * - `false` to block it (let the browser decide, usually moves to body).
-   *   - Notice this might break keyboard accessibility and should be used with caution.
-   * - An HTMLElement to redirect focus to instead of normal flow.
+   * - `true` to let focus-lock **manage** this element (keep it within the modal's focus trap).
+   * - `false` to let focus-lock **ignore** this element (allow focus to move outside the modal).
+   * - An HTMLElement to redirect focus to that element instead.
    * - Any other value (e.g., null, undefined) would act as `false`.
+   *
+   * @example
```
Comment (rivka-ungar) [inline packages/core/src/components/Modal/Modal/Modal.types.tsx:73]: Maybe consider creating a separate comprehensive doc with the use cases and examples? (Same like we had for the old dropdown)

File: packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { useCallback, useMemo } from "react";
```
Comment (rivka-ungar) [inline packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts:1]: Worth adding tests in later pr

File: packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts (hunk 1 above)
Comment (YossiSaadi) [reply packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts:22]: I want in that case to add to the `elementList`, not the `selectorList`

File: packages/core/src/components/Modal/Modal/Modal.types.tsx (hunk 2 above)
Comment (YossiSaadi) [reply packages/core/src/components/Modal/Modal/Modal.types.tsx:73]: I agree. what do you think about including it in the same story? or do you prefer having it in a total separate one?

File: packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts (hunk 3 above)
Comment (YossiSaadi) [reply packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts:1]: Done!

File: packages/core/src/components/Modal/Modal/Modal.types.tsx (hunk 2 above)
Comment (YossiSaadi) [reply packages/core/src/components/Modal/Modal/Modal.types.tsx:73]: https://monday.monday.com/boards/3532714909/pulses/18284854154

File: packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts (hunk 3 above)
Comment (rivka-ungar) [reply packages/core/src/components/Modal/hooks/useFocusEscapeTargets/useFocusEscapeTargets.ts:1]: 💪
